<!DOCTYPE html> <html lang="en"> <head> <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"> <meta charset="utf-8"> <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"> <meta http-equiv="X-UA-Compatible" content="IE=edge"> <title>On the importance of assessing the results of computational tools with ground truth data | Braulio Valdebenito-Maturana</title> <meta name="author" content="Braulio Valdebenito-Maturana"> <meta name="description" content="A fully provenanced example on validating the results of a computational tool"> <meta name="keywords" content="jekyll, jekyll-theme, academic-website, portfolio-website"> <link rel="stylesheet" href="/assets/css/bootstrap.min.css?a4b3f509e79c54a512b890d73235ef04"> <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/css/mdb.min.css" integrity="sha256-jpjYvU3G3N6nrrBwXJoVEYI/0zw8htfFnhT9ljN3JJw=" crossorigin="anonymous"> <link defer rel="stylesheet" href="https://unpkg.com/bootstrap-table@1.22.1/dist/bootstrap-table.min.css"> <link rel="stylesheet" href="/assets/css/academicons.min.css?f0b7046b84e425c55f3463ac249818f5"> <link rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700|Roboto+Slab:100,300,400,500,700|Material+Icons"> <link rel="stylesheet" href="/assets/css/jekyll-pygments-themes-github.css?19f3075a2d19613090fe9e16b564e1fe" media="" id="highlight_theme_light"> <link rel="shortcut icon" href="data:image/svg+xml,&lt;svg%20xmlns=%22http://www.w3.org/2000/svg%22%20viewBox=%220%200%20100%20100%22&gt;&lt;text%20y=%22.9em%22%20font-size=%2290%22&gt;%E2%9A%9B%EF%B8%8F&lt;/text&gt;&lt;/svg&gt;"> <link rel="stylesheet" href="/assets/css/main.css?d41d8cd98f00b204e9800998ecf8427e"> <link rel="canonical" href="https://bvaldebenitom.github.io/blog/2025/groundtruth/"> <link rel="stylesheet" href="/assets/css/jekyll-pygments-themes-native.css?e74e74bf055e5729d44a7d031a5ca6a5" media="none" id="highlight_theme_dark"> <script src="/assets/js/theme.js?96d6b3e1c3604aca8b6134c7afdd5db6"></script> <script src="/assets/js/dark_mode.js?9b17307bb950ffa2e34be0227f53558f"></script> </head> <body class="fixed-top-nav "> <header> <nav id="navbar" class="navbar navbar-light navbar-expand-sm fixed-top"> <div class="container"> <a class="navbar-brand title font-weight-lighter" href="/"><span class="font-weight-bold">Braulio </span>Valdebenito-Maturana</a> <button class="navbar-toggler collapsed ml-auto" type="button" data-toggle="collapse" data-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar top-bar"></span> <span class="icon-bar middle-bar"></span> <span class="icon-bar bottom-bar"></span> </button> <div class="collapse navbar-collapse text-right" id="navbarNav"> <ul class="navbar-nav ml-auto flex-nowrap"> <li class="nav-item "> <a class="nav-link" href="/">about</a> </li> <li class="nav-item "> <a class="nav-link" href="/blog/">blog</a> </li> <li class="nav-item "> <a class="nav-link" href="/publications/">publications</a> </li> <li class="nav-item "> <a class="nav-link" href="/repositories/">repositories</a> </li> <li class="toggle-container"> <button id="light-toggle" title="Change theme"> <i class="fa-solid fa-moon"></i> <i class="fa-solid fa-sun"></i> </button> </li> </ul> </div> </div> </nav> <progress id="progress" value="0"> <div class="progress-container"> <span class="progress-bar"></span> </div> </progress> </header> <div class="container mt-5"> <div class="post"> <header class="post-header"> <h1 class="post-title">On the importance of assessing the results of computational tools with ground truth data</h1> <p class="post-meta">October 17, 2025</p> <p class="post-tags"> <a href="/blog/2025"> <i class="fa-solid fa-calendar fa-sm"></i> 2025 </a>   ·   <a href="/blog/tag/rnaseq"> <i class="fa-solid fa-hashtag fa-sm"></i> rnaseq</a>   <a href="/blog/tag/transposableelements"> <i class="fa-solid fa-hashtag fa-sm"></i> transposableelements</a>     ·   <a href="/blog/category/reviews"> <i class="fa-solid fa-tag fa-sm"></i> reviews</a>   </p> </header> <article class="post-content"> <div id="table-of-contents"> <ul id="toc" class="section-nav"> <li class="toc-entry toc-h1"><a href="#preliminaries">PRELIMINARIES</a></li> <li class="toc-entry toc-h1"><a href="#what-you-will-need">WHAT YOU WILL NEED</a></li> <li class="toc-entry toc-h1"><a href="#step-1---genome-index">STEP 1 - GENOME INDEX</a></li> <li class="toc-entry toc-h1"><a href="#step-2---experiment-sequences">STEP 2 - EXPERIMENT SEQUENCES</a></li> <li class="toc-entry toc-h1"><a href="#step-3---read-simulation">STEP 3 - READ SIMULATION</a></li> <li class="toc-entry toc-h1"><a href="#step-4---setting-up-squire">STEP 4 - SETTING UP SQUIRE</a></li> <li class="toc-entry toc-h1"><a href="#step-5---mapping">STEP 5 - MAPPING</a></li> <li class="toc-entry toc-h1"><a href="#step-6---te-quantification-and-validation">STEP 6 - TE QUANTIFICATION AND VALIDATION</a></li> </ul> </div> <hr> <div id="markdown-content"> <p><strong>This is a supplementary material for my upcoming book chapter “Transposable Element analysis in OMICS data”</strong></p> <h1 id="preliminaries">PRELIMINARIES</h1> <p><a href="https://bvaldebenitom.github.io/blog/2024/igv_coverage/">In a previous post</a>, I briefly mentioned how the tool SQuIRE, which can be used for locus-specific measurements of Transposable Element (TE) expression from RNA-Seq data, can results in false positives. Here, I will expand on that, by showing a fully reproducible example of this. This post is not intended to discredit SQuIRE, but rather to show that for any computational tool we always need to think on ways to test their results with ground truth data, that is, with inputs that we know for a fact that should result in a specific output. In this line, here I designed a simple experiment to show how this can be achieved.</p> <h1 id="what-you-will-need">WHAT YOU WILL NEED</h1> <ul> <li>A STAR-indexed genome</li> <li>Experiment sequences</li> <li>ART read simulator</li> <li>SQuIRE</li> <li>IGV</li> </ul> <h1 id="step-1---genome-index">STEP 1 - GENOME INDEX</h1> <p>The first step is to get the genome FASTA file and GTF gene annotation file:</p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>wget http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz
wget http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/genes/hg38.ncbiRefSeq.gtf.gz

<span class="nb">gzip</span> <span class="nt">-d</span> hg38.fa.gz
<span class="nb">gzip</span> <span class="nt">-d</span> hg38.ncbiRefSeq.gtf.gz
</code></pre></div></div> <p>We can then build the index with STAR as follows:</p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>STAR   <span class="nt">--runMode</span> genomeGenerate      <span class="nt">--runThreadN</span> 21   <span class="nt">--genomeDir</span> hg38_STAR_INDEX  <span class="nt">--genomeFastaFiles</span> hg38.fa      <span class="nt">--sjdbGTFfile</span> hg38.ncbiRefSeq.gtf
</code></pre></div></div> <h1 id="step-2---experiment-sequences">STEP 2 - EXPERIMENT SEQUENCES</h1> <p>For this simple experiment, in the <code class="language-plaintext highlighter-rouge">experiment_sequences.fa</code> file, we only have two sequences:</p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">grep</span> <span class="s2">"^&gt;"</span> experiment_sequences.fa
</code></pre></div></div> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## &gt;chr4|4013499|4019703|L1PA4:L1:LINE|5.8|-::chr4:4013499-4019703(-)
## &gt;NM_020364.4 Homo sapiens deleted in azoospermia 3 (DAZ3), mRNA
</code></pre></div></div> <p>One is a sequence from a TE <strong>located in chromosome 4</strong>, while the other is from the protein-coding sequence of the DAZ3 gene <strong>located in chromosome Y</strong>.</p> <h1 id="step-3---read-simulation">STEP 3 - READ SIMULATION</h1> <p>We can get the ART read simulator from <a href="https://www.niehs.nih.gov/research/resources/software/biostatistics/art" rel="external nofollow noopener" target="_blank">from its official website</a>:</p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>wget https://www.niehs.nih.gov/sites/default/files/2024-02/artbinmountrainier2016.06.05linux64.tgz
<span class="nb">tar</span> <span class="nt">-xvf</span> artbinmountrainier2016.06.05linux64.tgz 
</code></pre></div></div> <p>To perform read simulation, we set then <code class="language-plaintext highlighter-rouge">-ss HS25</code> for single-end layout, <code class="language-plaintext highlighter-rouge">-l 150</code> for read length of 150 bp, <code class="language-plaintext highlighter-rouge">-f 30</code> to simulate reads representing 30X coverage of the sequences in <code class="language-plaintext highlighter-rouge">-i experiment_sequences.fa</code>, and <code class="language-plaintext highlighter-rouge">-o experiment_sequences</code> as output prefix.</p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>art_bin_MountRainier/art_illumina <span class="nt">-ss</span> HS25 <span class="nt">-i</span> experiment_sequences.fa <span class="nt">-l</span> 150 <span class="nt">-f</span> 30 <span class="nt">-o</span> experiment_sequences
</code></pre></div></div> <p>This will give us the <code class="language-plaintext highlighter-rouge">experiment_sequences.fq</code> file, which has the FASTQ simulated reads to map against the human genome.</p> <h1 id="step-4---setting-up-squire">STEP 4 - SETTING UP SQUIRE</h1> <p>The installation of SQuIRE is a relatively <em>seamless</em> process if you already have Conda. Their repository is well documented and provides instruction to install the tool from scratch, including setting up Conda. You can check their instructions <a href="https://github.com/wyang17/SQuIRE/" rel="external nofollow noopener" target="_blank">here</a></p> <p>Once installed, we need to check that the <code class="language-plaintext highlighter-rouge">Count</code> tool is available:</p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>squire Count <span class="nt">-h</span>

usage: squire Count <span class="o">[</span><span class="nt">-h</span><span class="o">]</span> <span class="o">[</span><span class="nt">-m</span> &lt;folder&gt;] <span class="o">[</span><span class="nt">-c</span> &lt;folder&gt;] <span class="o">[</span><span class="nt">-o</span> &lt;folder&gt;]
                    <span class="o">[</span><span class="nt">-t</span> &lt;folder&gt;] <span class="o">[</span><span class="nt">-f</span> &lt;folder&gt;] <span class="nt">-r</span> &lt;int&gt; <span class="o">[</span><span class="nt">-n</span> &lt;str&gt;]
                    <span class="o">[</span><span class="nt">-b</span> &lt;build&gt;] <span class="o">[</span><span class="nt">-p</span> &lt;int&gt;] <span class="o">[</span><span class="nt">-s</span> &lt;int&gt;] <span class="o">[</span><span class="nt">-e</span> EM] <span class="o">[</span><span class="nt">-v</span><span class="o">]</span>

Arguments:
  <span class="nt">-h</span>, <span class="nt">--help</span>            show this <span class="nb">help </span>message and <span class="nb">exit</span>
  <span class="nt">-m</span> &lt;folder&gt;, <span class="nt">--map_folder</span> &lt;folder&gt;
                        Folder location of outputs from SQuIRE Map <span class="o">(</span>optional,
                        default <span class="o">=</span> <span class="s1">'squire_map'</span><span class="o">)</span>
  <span class="nt">-c</span> &lt;folder&gt;, <span class="nt">--clean_folder</span> &lt;folder&gt;
                        Folder location of outputs from SQuIRE Clean
                        <span class="o">(</span>optional, default <span class="o">=</span> <span class="s1">'squire_clean'</span><span class="o">)</span>
  <span class="nt">-o</span> &lt;folder&gt;, <span class="nt">--count_folder</span> &lt;folder&gt;
                        Destination folder <span class="k">for </span>output files<span class="o">(</span>optional, default
                        <span class="o">=</span> <span class="s1">'squire_count'</span><span class="o">)</span>
  <span class="nt">-t</span> &lt;folder&gt;, <span class="nt">--tempfolder</span> &lt;folder&gt;
                        Folder <span class="k">for </span>tempfiles <span class="o">(</span>optional<span class="p">;</span> <span class="nv">default</span><span class="o">=</span><span class="s1">'count_folder'</span><span class="o">)</span>
  <span class="nt">-f</span> &lt;folder&gt;, <span class="nt">--fetch_folder</span> &lt;folder&gt;
                        Folder location of outputs from SQuIRE Fetch
                        <span class="o">(</span>optional, default <span class="o">=</span> <span class="s1">'squire_fetch)'</span>
  <span class="nt">-r</span> &lt;int&gt;, <span class="nt">--read_length</span> &lt;int&gt;
                        Read length <span class="o">(</span><span class="k">if </span>trim3 selected, after trimming<span class="p">;</span>
                        required<span class="o">)</span><span class="nb">.</span>
  <span class="nt">-n</span> &lt;str&gt;, <span class="nt">--name</span> &lt;str&gt;
                        Common <span class="nb">basename </span><span class="k">for </span>input files <span class="o">(</span>required <span class="k">if </span>more than
                        one bam file <span class="k">in </span>map_folder<span class="o">)</span>
  <span class="nt">-b</span> &lt;build&gt;, <span class="nt">--build</span> &lt;build&gt;
                        UCSC designation <span class="k">for </span>genome build, eg. <span class="s1">'hg38'</span>
                        <span class="o">(</span>required <span class="k">if </span>more than 1 build <span class="k">in </span>clean_folder<span class="o">)</span>
  <span class="nt">-p</span> &lt;int&gt;, <span class="nt">--pthreads</span> &lt;int&gt;
                        Launch &lt;int&gt; parallel threads<span class="o">(</span>optional<span class="p">;</span> <span class="nv">default</span><span class="o">=</span><span class="s1">'1'</span><span class="o">)</span>
  <span class="nt">-s</span> &lt;int&gt;, <span class="nt">--strandedness</span> &lt;int&gt;
                        <span class="s1">'0'</span> <span class="k">if </span>unstranded eg Standard Illumina, 1 <span class="k">if </span>first-
                        strand eg Illumina Truseq, dUTP, NSR, NNSR, 2 <span class="k">if
                        </span>second-strand, eg Ligation, Standard SOLiD
                        <span class="o">(</span>optional,default<span class="o">=</span>0<span class="o">)</span>
  <span class="nt">-e</span> EM, <span class="nt">--EM</span> EM        Run estimation-maximization on TE counts given number
                        of <span class="nb">times</span> <span class="o">(</span>optional, specify 0 <span class="k">if </span>no EM desired<span class="p">;</span>
                        <span class="nv">default</span><span class="o">=</span>auto<span class="o">)</span>
  <span class="nt">-v</span>, <span class="nt">--verbosity</span>       Want messages and runtime printed to stderr <span class="o">(</span>optional<span class="p">;</span>
                        <span class="nv">default</span><span class="o">=</span>False<span class="o">)</span>

</code></pre></div></div> <p>We now need to run these commands to prepare some additional files and directories required for the quantification of TE expression:</p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>squire Fetch <span class="nt">-b</span> hg38 <span class="nt">-c</span> <span class="nt">-g</span> <span class="nt">-r</span> <span class="nt">-v</span>
squire Clean <span class="nt">-r</span> squire_fetch/hg38_rmsk.txt <span class="nt">-v</span>
</code></pre></div></div> <p>Now, we should have the directories <code class="language-plaintext highlighter-rouge">squire_fetch</code> and <code class="language-plaintext highlighter-rouge">squire_clean</code>:</p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>tree squire_fetch
tree squire_clean
</code></pre></div></div> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## squire_fetch
## ├── hg38chromFa.tar.gz
## ├── hg38_chromInfo.txt
## ├── hg38_refGene.bed
## ├── hg38_refGene.genepred
## ├── hg38_refGene.gtf
## └── hg38_rmsk.txt
## 
## 0 directories, 6 files
## squire_clean
## ├── hg38_all.bed
## └── hg38_all_copies.txt
## 
## 0 directories, 2 files
</code></pre></div></div> <h1 id="step-5---mapping">STEP 5 - MAPPING</h1> <p>Now we are all set!</p> <p>We can map our simulated sequences <code class="language-plaintext highlighter-rouge">experiment_sequences.fq</code> against the genome index, using the same options that <code class="language-plaintext highlighter-rouge">squire Map</code> <a href="https://github.com/wyang17/SQuIRE/blob/master/squire/Map.py#L173" rel="external nofollow noopener" target="_blank">uses</a></p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>STAR <span class="nt">--runThreadN</span> 1 <span class="nt">--clip3pNbases</span> 0 <span class="nt">--outFilterMultimapNmax</span> 100 <span class="nt">--winAnchorMultimapNmax</span> 100 <span class="nt">--genomeDir</span> hg38_STAR_INDEX <span class="nt">--readFilesIn</span> experiment_sequences.fq <span class="nt">--outFileNamePrefix</span> squire_map/experiment_alignment_ <span class="nt">--outSAMtype</span> BAM SortedByCoordinate <span class="nt">--outSAMattributes</span> All <span class="nt">--outSAMstrandField</span> intronMotif <span class="nt">--outSAMattrIHstart</span> 0 <span class="nt">--sjdbGTFfile</span> hg38.ncbiRefSeq.gtf <span class="nt">--twopassMode</span> Basic
</code></pre></div></div> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>tree squire_map
</code></pre></div></div> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## squire_map
## ├── experiment_alignment_Aligned.sortedByCoord.out.bam
## ├── experiment_alignment_Aligned.sortedByCoord.out.bam.bai
## ├── experiment_alignment_Log.final.out
## ├── experiment_alignment_Log.out
## ├── experiment_alignment_Log.progress.out
## ├── experiment_alignment_SJ.out.tab
## ├── experiment_alignment__STARgenome
## │   ├── exonGeTrInfo.tab
## │   ├── exonInfo.tab
## │   ├── geneInfo.tab
## │   ├── sjdbInfo.txt
## │   ├── sjdbList.fromGTF.out.tab
## │   ├── sjdbList.out.tab
## │   └── transcriptInfo.tab
## └── experiment_alignment__STARpass1
##     ├── Log.final.out
##     └── SJ.out.tab
## 
## 2 directories, 15 files
</code></pre></div></div> <h1 id="step-6---te-quantification-and-validation">STEP 6 - TE QUANTIFICATION AND VALIDATION</h1> <p>With the mapping result, we can run the <code class="language-plaintext highlighter-rouge">Count</code> quantification tool:</p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>squire Count <span class="nt">-m</span> squire_map/ <span class="nt">-r</span> 150 <span class="nt">-c</span> squire_clean/ <span class="nt">-v</span>
</code></pre></div></div> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>tree squire_count
</code></pre></div></div> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## squire_count
## ├── experiment_alignment_Aligned.sortedByCoord.out_abund.txt
## ├── experiment_alignment_Aligned.sortedByCoord.out.gtf
## ├── experiment_alignment_Aligned.sortedByCoord.out_refGenecounts.txt
## ├── experiment_alignment_Aligned.sortedByCoord.out_subFcounts.txt
## └── experiment_alignment_Aligned.sortedByCoord.out_TEcounts.txt
## 
## 0 directories, 5 files
</code></pre></div></div> <p>Of these files, we are interested in <code class="language-plaintext highlighter-rouge">experiment_alignment_Aligned.sortedByCoord.out_TEcounts.txt</code>, which provides the quantification of TEs. We can check the most important columns:</p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">awk</span> <span class="s1">'BEGIN{FS=OFS="\t"}{print $1,$2,$3,$4,$15,$16,$17}'</span> squire_count/experiment_alignment_Aligned.sortedByCoord.out_TEcounts.txt
</code></pre></div></div> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## tx_chr	tx_start	tx_stop	TE_ID	uniq_counts	tot_counts	tot_reads
## chr4	4013502	4019680	chr4|4013498|4019703|L1PA4:L1:LINE|58|-	1118	1183.90	1216
## chr8	96955960	96956120	chr8|96955434|96956163|L1PA4:L1:LINE|47|+	0	4.31	6
## chr10	22061677	22061836	chr10|22060724|22066859|L1PA5:L1:LINE|53|+	0	4.20	6
## chr4	9617631	9617845	chr4|9617608|9619322|L1PA4:L1:LINE|86|+	0	5.29	15
## chr10	82344563	82344736	chr10|82344062|82350086|L1PA2:L1:LINE|25|-	0	3.79	12
## chr3	112659647	112659820	chr3|112659127|112665164|L1PA3:L1:LINE|29|-	0	3.79	12
## chr5	62656507	62656735	chr5|62655618|62660602|L1P1:L1:LINE|57|+	0	3.34	14
## chr5	93766091	93766242	chr5|93761070|93767199|L1PA4:L1:LINE|55|+	0	1.73	2
## chr2	133466941	133467107	chr2|133465007|133471038|L1PA2:L1:LINE|29|-	0	1.53	5
## chr20	29366629	29366779	chr20|29366059|29372220|L1PA4:L1:LINE|54|-	1	1.00	1
## chr5	114891825	114891975	chr5|114891256|114895798|L1PA4:L1:LINE|49|-	1	1.00	1
## chrX	97578708	97578858	chrX|97578170|97584188|L1PA3:L1:LINE|25|-	1	1.00	1
## chrX	75693308	75693519	chrX|75690170|75694093|L1PA4:L1:LINE|58|-	0	1.24	15
## chr3	115315428	115315678	chr3|115312305|115318470|L1PA3:L1:LINE|34|-	0	1.45	22
## chr3	121455034	121455284	chr3|121452690|121458395|L1PA3:L1:LINE|39|+	0	1.45	22
## chr12	29982525	29982675	chr12|29981945|29982867|L1PA4:L1:LINE|29|-	0	0.50	1
## chr1	48118716	48118866	chr1|48118135|48120412|L1PA4:L1:LINE|31|-	0	0.50	1
## chrY	24773569	24784208	chrY|24776006|24782046|L1PA2:L1:LINE|31|+	15	15.00	15
</code></pre></div></div> <p><strong>Wait a minute!!</strong>. So, from our experiment design, we should only expect TE expression at chromosome 4, specifically at <code class="language-plaintext highlighter-rouge">chr4:4013499-4019703</code>. We indeed see this on the first row, but there is expression detected at other chromosomes:</p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">awk</span> <span class="s1">'BEGIN{FS=OFS="\t"}{print $1}'</span> squire_count/experiment_alignment_Aligned.sortedByCoord.out_TEcounts.txt|sort|uniq <span class="nt">-c</span>
</code></pre></div></div> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>##       1 chr1
##       2 chr10
##       1 chr12
##       1 chr2
##       1 chr20
##       3 chr3
##       2 chr4
##       3 chr5
##       1 chr8
##       2 chrX
##       1 chrY
##       1 tx_chr
</code></pre></div></div> <p>We see a result in chromosome Y, at locations 24773569 to 24784208:</p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">awk</span> <span class="s1">'BEGIN{FS=OFS="\t"}{print $1,$2,$3,$4}'</span> squire_count/experiment_alignment_Aligned.sortedByCoord.out_TEcounts.txt|grep <span class="s2">"chrY"</span>
</code></pre></div></div> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## chrY	24773569	24784208	chrY|24776006|24782046|L1PA2:L1:LINE|31|+
</code></pre></div></div> <p>This regions overlap with the location of DAZ3, approximately at <code class="language-plaintext highlighter-rouge">chrY:24761069-24815492</code>, which <strong>is the gene from which we also simulated sequences</strong>.</p> <p>We can now inspect these results in IGV. First, we create a <em>bedgraph</em> representation of our BAM file using the <code class="language-plaintext highlighter-rouge">bamCoverage</code> tool from <a href="https://deeptools.readthedocs.io/en/latest/" rel="external nofollow noopener" target="_blank">deepTools</a>:</p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>bamCoverage <span class="nt">--bam</span> squire_map/experiment_alignment_Aligned.sortedByCoord.out.bam <span class="nt">--outFileName</span> experiment.bedgraph <span class="nt">--outFileFormat</span> bedgraph <span class="nt">--binSize</span> 5 <span class="nt">--region</span> chrY:24761069:24815492
</code></pre></div></div> <p>A dummy representation of SQuIRE’s results as <em>bedgraph</em> can be created as follows:</p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">echo</span> <span class="nt">-e</span> <span class="s2">"chrY</span><span class="se">\t</span><span class="s2">24773569</span><span class="se">\t</span><span class="s2">24784208</span><span class="se">\t</span><span class="s2">15"</span> <span class="o">&gt;</span> squire.bedgraph
</code></pre></div></div> <p>Those 2 files can be <em>seamlessly</em> loaded in IGV, and we will be able to see this:</p> <div class="row mt-3"> <div class="col-sm mt-3 mt-md-0"> <figure> <picture> <source class="responsive-img-srcset" media="(max-width: 480px)" srcset="/assets/img/03_squire/igv_snapshot_squire-480.webp"></source> <source class="responsive-img-srcset" media="(max-width: 800px)" srcset="/assets/img/03_squire/igv_snapshot_squire-800.webp"></source> <source class="responsive-img-srcset" media="(max-width: 1400px)" srcset="/assets/img/03_squire/igv_snapshot_squire-1400.webp"></source> <img src="/assets/img/03_squire/igv_snapshot_squire.png" class="img-fluid rounded z-depth-1" width="auto" height="auto" data-zoomable="" onerror="this.onerror=null; $('.responsive-img-srcset').remove();"> </picture> </figure> </div> </div> <div class="caption"> Genomic snapshot of the region chrY:24761069-24815492 </div> <p>If proficient with <a href="https://github.com/showteeth/ggcoverage/" rel="external nofollow noopener" target="_blank">ggCoverage</a>, we can create a similar representation of the IGV result, as a fully <em>vectorial</em> image:</p> <div class="row mt-3"> <div class="col-sm mt-3 mt-md-0"> <figure> <picture> <source class="responsive-img-srcset" media="(max-width: 480px)" srcset="/assets/img/03_squire/ggcoverage_squire-480.webp"></source> <source class="responsive-img-srcset" media="(max-width: 800px)" srcset="/assets/img/03_squire/ggcoverage_squire-800.webp"></source> <source class="responsive-img-srcset" media="(max-width: 1400px)" srcset="/assets/img/03_squire/ggcoverage_squire-1400.webp"></source> <img src="/assets/img/03_squire/ggcoverage_squire.png" class="img-fluid rounded z-depth-1" width="auto" height="auto" data-zoomable="" onerror="this.onerror=null; $('.responsive-img-srcset').remove();"> </picture> </figure> </div> </div> <div class="caption"> Genomic snapshot of the region chrY:24761069-24815492 with ggCoverage. </div> <p>I found that SQuIRE over-estimates intronic TE expression due to <a href="https://github.com/wyang17/SQuIRE/blob/master/squire/Count.py#L269" rel="external nofollow noopener" target="_blank">indiscriminate use of BEDtools</a>. When there are <em>spliced</em> alignments, BEDtools by default will convert the interval in the SAM/BAM file to start position plus length of CIGAR string. As an example, for a CIGAR string like 20M3000N80N, and alignment position at 1000000, it will create an interval of length 20+3000+80 = 3100, and end position at 1003100. We need to either use <code class="language-plaintext highlighter-rouge">bedtools bamtobed</code> with the <code class="language-plaintext highlighter-rouge">-split</code> option to create an intermediate file (which we can further check that correctly has split intervals), or use the <code class="language-plaintext highlighter-rouge">-split</code> option in <code class="language-plaintext highlighter-rouge">bedtools intersect</code> so it does not artifically creates intervals longer than the sequencing reads.</p> <p>This is why whenever I use a tool, I design what I call “ground truth” experiments, in which with a simple test dataset, I can accurately tell if the tool does what it is supposed to do. Although it is very exciting that nowadays bioinformatics tool become more readily available, we should always exercise caution in their use, as this can result in arriving at biological conclusions that might not be entirely true. For example, using SQuIRE, a previous article made a point of “intron retention” affecting TE quantification. From my experience, and the experiment shown here, I could argue that there migh t be over-estimation of intronic TE expression, but that does not correspond to actual “intron retention”.</p> </div> </article> <br> <hr> <br> <ul class="list-disc pl-8"></ul> <h2 class="text-3xl font-semibold mb-4 mt-12">Enjoy Reading This Article?</h2> <p class="mb-2">Here are some more articles you might like to read next:</p> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2024/nextflow_vs_snakemake/">Nextflow vs Snakemake</a> </li> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2024/igv_coverage/">Automatization of IGV snapshots</a> </li> </div> </div> <footer class="fixed-bottom"> <div class="container mt-0"> © Copyright 2025 Braulio Valdebenito-Maturana. Powered by <a href="https://jekyllrb.com/" target="_blank" rel="external nofollow noopener">Jekyll</a> with <a href="https://github.com/alshedivat/al-folio" rel="external nofollow noopener" target="_blank">al-folio</a> theme. Hosted by <a href="https://pages.github.com/" target="_blank" rel="external nofollow noopener">GitHub Pages</a>. Photos from <a href="https://unsplash.com" target="_blank" rel="external nofollow noopener">Unsplash</a>. </div> </footer> <script src="https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.min.js" integrity="sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=" crossorigin="anonymous"></script> <script src="/assets/js/bootstrap.bundle.min.js"></script> <script src="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/js/mdb.min.js" integrity="sha256-NdbiivsvWt7VYCt6hYNT3h/th9vSTL4EDWeGs5SN3DA=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/masonry-layout@4.2.2/dist/masonry.pkgd.min.js" integrity="sha256-Nn1q/fx0H7SNLZMQ5Hw5JLaTRZp0yILA/FRexe19VdI=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/imagesloaded@4/imagesloaded.pkgd.min.js"></script> <script defer src="/assets/js/masonry.js" type="text/javascript"></script> <script defer src="https://cdn.jsdelivr.net/npm/medium-zoom@1.0.8/dist/medium-zoom.min.js" integrity="sha256-7PhEpEWEW0XXQ0k6kQrPKwuoIomz8R8IYyuU1Qew4P8=" crossorigin="anonymous"></script> <script defer src="/assets/js/zoom.js?7b30caa5023af4af8408a472dc4e1ebb"></script> <script defer src="https://unpkg.com/bootstrap-table@1.22.1/dist/bootstrap-table.min.js"></script> <script src="/assets/js/no_defer.js?d633890033921b33e0ceb13d22340a9c"></script> <script defer src="/assets/js/common.js?acdb9690d7641b2f8d40529018c71a01"></script> <script defer src="/assets/js/copy_code.js?07b8786bab9b4abe90d10e61f7d12ff7" type="text/javascript"></script> <script async src="https://d1bxh8uas1mnw7.cloudfront.net/assets/embed.js"></script> <script async src="https://badge.dimensions.ai/badge.js"></script> <script type="text/javascript">window.MathJax={tex:{tags:"ams"}};</script> <script defer type="text/javascript" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3.2.0/es5/tex-mml-chtml.js"></script> <script defer src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script type="text/javascript">function progressBarSetup(){"max"in document.createElement("progress")?(initializeProgressElement(),$(document).on("scroll",function(){progressBar.attr({value:getCurrentScrollPosition()})}),$(window).on("resize",initializeProgressElement)):(resizeProgressBar(),$(document).on("scroll",resizeProgressBar),$(window).on("resize",resizeProgressBar))}function getCurrentScrollPosition(){return $(window).scrollTop()}function initializeProgressElement(){let e=$("#navbar").outerHeight(!0);$("body").css({"padding-top":e}),$("progress-container").css({"padding-top":e}),progressBar.css({top:e}),progressBar.attr({max:getDistanceToScroll(),value:getCurrentScrollPosition()})}function getDistanceToScroll(){return $(document).height()-$(window).height()}function resizeProgressBar(){progressBar.css({width:getWidthPercentage()+"%"})}function getWidthPercentage(){return getCurrentScrollPosition()/getDistanceToScroll()*100}const progressBar=$("#progress");window.onload=function(){setTimeout(progressBarSetup,50)};</script> </body> </html>